### PR TITLE
txn: Snapshot can set `SortedCustomRetrievers` for custom data retrieving

### LINKED ISSUE #27587: Add method for snapshot for setting custom range retrievers

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #27587

Allow settings custom retrievers for snapshot. Is is required for refactor of temporary table codes.

[Here](https://github.com/pingcap/tidb/pull/26948/files#diff-78a290ecfa9ee38aa2297340f8ab514c2f8390eaa66ce186a53ec1110f1d3ad1R59-R67) is the draft using this api to custom temporary read.

### What is changed and how it works?

We added an option `kv. SortedCustomRetrievers` to set custom retrievers for snapshot. After it, if the the request is located in the custom range, the snapshot will fetch data from retrievers.

Suppose we want to read a temporary table data from memory. A simple demo code below:

```
// var (
//      store kv.Storage
//      ver kv.Version
//      tbID int64 // tbID is the temprorary table's ID
// )
// init store, ver and tbID...

// Create a retriever. Its range is temporary table's range
retriever := txn.NewRangeRetriever(
    tablecodec.EncodeTablePrefix(tblID), 
    tablecodec.EncodeTablePrefix(tblID+1), 
    CreateNewMemBuffer()
)

snapshot := store.GetSnapshot(kv.ver)

// Set the retriever to snapshot
snapshot.SetOption(kv.SortedCustomRetrievers, []*txn.RangedKVRetriever{retriever})

// After that when getting the above temporary table keys, it will read the retriever
rowKey := tablecodec.EncodeRowKeyWithHandle(tblID, kv.IntHandle(1))
snapshot.Get(context.TODO(), rowKey)
```

### Check List 

Tests 

- [x] Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by disksing — APPROVED** [review]

File: store/driver/txn/snapshot.go [hunk 1]
```diff
@@ -28,28 +28,66 @@ import (
...
 // NewSnapshot creates a kv.Snapshot with txnsnapshot.KVSnapshot.
 func NewSnapshot(snapshot *txnsnapshot.KVSnapshot) kv.Snapshot {
-	return &tikvSnapshot{snapshot}
+	return &tikvSnapshot{snapshot, nil}
 }
 
 // BatchGet gets all the keys' value from kv-server and returns a map contains key/value pairs.
 // The map will not contain nonexistent keys.
 func (s *tikvSnapshot) BatchGet(ctx context.Context, keys []kv.Key) (map[string][]byte, error) {
+	customData := make(map[string][]byte)
```
Comment (tiancaiamao) [inline store/driver/txn/snapshot.go:43]: `if s.customRetrievers != nil` ?

Comment (tiancaiamao) [reply store/driver/txn/snapshot.go:43]: This code path is performance sensitive, avoid any unnecessary allocation please.

**Review by tiancaiamao — COMMENTED** [review]
rest LGTM

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**